framework/src/meshmodifiers/RenameBlock.C

```diff
@@ -78,30 +80,76 @@ RenameBlock::modify()
   // have been put into the mesh yet, eg old_block_name might have been inserted by
   // another MeshModifier
   if (isParamValid("old_block_id"))
+  {
+    // suppied old_block_id
     _old_block_id = getParam<std::vector<SubdomainID>>("old_block_id");
-  else // must have supplied old_block_name
-    _old_block_id =
-        _mesh_ptr->getSubdomainIDs(getParam<std::vector<SubdomainName>>("old_block_name"));
+    _old_block_name.resize(0);
+    for (const auto & block_id : _old_block_id)
+      _old_block_name.push_back(_mesh_ptr->getMesh().subdomain_name(block_id));
+  }
+  else
+  {
+    // must have supplied old_block_name
+    _old_block_name = getParam<std::vector<SubdomainName>>("old_block_name");
+    _old_block_id = _mesh_ptr->getSubdomainIDs(_old_block_name);
+  }
 
+  // construct new_block_id and new_block_name
   if (isParamValid("new_block_id"))
   {
     _new_block_id = getParam<std::vector<SubdomainID>>("new_block_id");
     if (_new_block_id.size() != _old_block_id.size())
       mooseError("RenameBlock: The vector of old_block information must have the same length as "
                  "the vector of new_block information\n");
-    for (const auto & elem : _mesh_ptr->getMesh().active_element_ptr_range())
-      for (unsigned i = 0; i < _old_block_id.size(); ++i)
-        if (elem->subdomain_id() == _old_block_id[i])
-          elem->subdomain_id() = _new_block_id[i];
+
+    // construct the _new_block_name.  Mostly this will just be
+    // _old_block_name, but in cases like
+    // old_block_id = '1 2 3'
+    // new_block_id = '1 1 1'
+    // the _new_block_name will be 'block1_name block1_name block1_name'
+    _new_block_name.resize(_new_block_id.size());
+    std::map<SubdomainID, SubdomainName> new_block_id_to_name;
+    for (unsigned i = 0; i < _new_block_id.size(); ++i)
+    {
+      // note that if _new_block_id[i] has already been encountered
+      // (such as in the above example) the following insert will
+      // do nothing.  Ie, we only put the _old_block_name from the
+      // first occurance of new_block_id into new_block_id_to_name
+      new_block_id_to_name.insert(
+          std::pair<SubdomainID, SubdomainName>(_new_block_id[i], _old_block_name[i]));
+      _new_block_name[i] = new_block_id_to_name[_new_block_id[i]];
+    }
```

Comment:
Hmm, I find this code a little confusing. You are building a map only to use it (through indirection) in the very next statement. It works, but I think just making a local subroutine returns a name, given an ID would be cleaner and would allow you to drop this whole temporary map business.

Reply:
Thanks.  It looks better now.